File: examples/python/part-8/24-end-to-end-secure-agent-go.py
# Illustrative examples for notes/part-8-practice/24-end-to-end-secure-agent-go.md
# Not for production use. Licensed under MIT (see LICENSE-CODE).

from __future__ import annotations

import re
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Any, Callable, Optional, Protocol, runtime_checkable


# --- Базовые типы ---


class RiskLevel(str, Enum):
    LOW = "Low"
    MEDIUM = "Medium"
    HIGH = "High"
    CRITICAL = "Critical"


class TrustLevel(str, Enum):
    TRUSTED_RUNTIME = "trusted_runtime"
    UNTRUSTED_USER = "untrusted_user"
    UNTRUSTED_TOOL = "untrusted_tool"


@dataclass
class ToolCall:
    name: str
    args: dict[str, Any]


@dataclass
class Plan:
    final_answer: str = ""
    tool_call: Optional[ToolCall] = None
    reason: str = ""


@dataclass
class ContextBlock:
    source: str
    trust: TrustLevel
    text: str


@dataclass
class RunRequest:
    run_id: str
    user_id: str
    task: str


@dataclass
class RunResult:
    run_id: str
    final_answer: str = ""
    tool_calls: list[ToolCall] = field(default_factory=list)
    blocked: list[str] = field(default_factory=list)


# --- Интерфейсы runtime ---


@runtime_checkable
class InputGuard(Protocol):
    def check(self, task: str) -> None: ...


@runtime_checkable
class ContextBuilder(Protocol):
    def build(self, req: RunRequest) -> list[ContextBlock]: ...


@runtime_checkable
class Planner(Protocol):
    def plan(self, req: RunRequest, blocks: list[ContextBlock]) -> Plan: ...


@runtime_checkable
class Policy(Protocol):
    def allow(self, req: RunRequest, call: ToolCall) -> RiskLevel: ...


@runtime_checkable
class SchemaValidator(Protocol):
    def validate(self, call: ToolCall) -> None: ...


@runtime_checkable
class ApprovalGate(Protocol):
    def approve(self, req: RunRequest, call: ToolCall, risk: RiskLevel) -> None: ...


@runtime_checkable
class ToolExecutor(Protocol):
    def execute(self, call: ToolCall) -> str: ...


@runtime_checkable
class OutputValidator(Protocol):
    def validate(self, text: str) -> None: ...


@runtime_checkable
class AuditLogger(Protocol):
    def log(self, event: "AuditEvent") -> None: ...


# --- Audit event ---


@dataclass
class AuditEvent:
    time: datetime
    run_id: str
    event: str
    component: str
    user_id: str = ""
    tool: str = ""
    risk: RiskLevel = RiskLevel.LOW
    decision: str = ""
    reason: str = ""
    attrs: dict[str, Any] = field(default_factory=dict)


# --- Run budget ---


@dataclass
class Budget:
    max_steps: int
    max_tool_calls: int
    max_denied: int
    steps: int = 0
    tool_calls: int = 0
    denied: int = 0

    def check(self) -> None:
        if self.steps > self.max_steps:
            raise RuntimeError("budget exceeded: max steps")
        if self.tool_calls > self.max_tool_calls:
            raise RuntimeError("budget exceeded: max tool calls")
        if self.denied > self.max_denied:
            raise RuntimeError("budget exceeded: max denied actions")


# --- Runtime агента ---


@dataclass
class Runtime:
    input_guard: InputGuard
    context_builder: ContextBuilder
    planner: Planner
    policy: Policy
    schema: SchemaValidator
    approval: ApprovalGate
    executor: ToolExecutor
    output: OutputValidator
    audit: AuditLogger
    budget_factory: Callable[[], Budget]

    def run(self, req: RunRequest) -> RunResult:
        if not req.run_id:
            raise ValueError("run_id is required")

        result = RunResult(run_id=req.run_id)
        budget = self.budget_factory()

        try:
            self.input_guard.check(req.task)
        except Exception as exc:
            self.audit.log(
                AuditEvent(
                    time=datetime.now(timezone.utc),
                    run_id=req.run_id,
                    user_id=req.user_id,
                    event="input_blocked",
                    component="input_guard",
                    decision="denied",
                    reason=str(exc),
                )
            )
            raise

        blocks = self.context_builder.build(req)

        while True:
            budget.steps += 1
            try:
                budget.check()
            except Exception as exc:
                result.blocked.append("budget_exceeded")
                raise exc

            plan = self.planner.plan(req, blocks)

            if plan.tool_call is None:
                try:
                    self.output.validate(plan.final_answer)
                except Exception as exc:
                    result.blocked.append("output_blocked")
                    raise exc

                result.final_answer = plan.final_answer
                return result

            call = plan.tool_call
            budget.tool_calls += 1

            try:
                risk = self.policy.allow(req, call)
            except Exception as exc:
                budget.denied += 1
                result.blocked.append(f"tool_denied:{call.name}")
                self.audit.log(
                    AuditEvent(
                        time=datetime.now(timezone.utc),
                        run_id=req.run_id,
                        user_id=req.user_id,
                        event="tool_denied",
                        component="policy",
                        tool=call.name,
                        decision="denied",
                        reason=str(exc),
                    )
                )
                raise

            try:
                self.schema.validate(call)
            except Exception as exc:
                budget.denied += 1
                result.blocked.append(f"schema_validation_failed:{call.name}")
                raise exc

            if requires_approval(risk):
                try:
                    self.approval.approve(req, call, risk)
                except Exception as exc:
                    budget.denied += 1
                    result.blocked.append(f"approval_rejected:{call.name}")
                    raise exc

            observation = self.executor.execute(call)
            result.tool_calls.append(call)

            blocks.append(
                ContextBlock(
                    source=f"tool:{call.name}",
                    trust=TrustLevel.UNTRUSTED_TOOL,
                    text=observation,
                )
            )


def requires_approval(risk: RiskLevel) -> bool:
    return risk in (RiskLevel.HIGH, RiskLevel.CRITICAL)


# --- Input guardrails ---


@dataclass
class Guard:
    max_chars: int

    def check(self, task: str) -> None:
        if not task.strip():
            raise ValueError("empty task")

        if len(task) > self.max_chars:
            raise ValueError("input too large")

        lower = task.lower()
        suspicious = [
            "ignore previous instructions",
            "disregard system prompt",
            "reveal your system prompt",
            "send secrets",
            "exfiltrate",
        ]

        for marker in suspicious:
            if marker in lower:
                raise ValueError("possible prompt injection")


# --- Context isolation ---


class Memory(Protocol):
    def read_for_user(self, user_id: str) -> list[ContextBlock]: ...


@dataclass
class Builder:
    memory: Memory

    def build(self, req: RunRequest) -> list[ContextBlock]:
        memory_blocks = self.memory.read_for_user(req.user_id)
        blocks = [
            ContextBlock(
                source="user_task",
                trust=TrustLevel.UNTRUSTED_USER,
                text=req.task,
            )
        ]
        blocks.extend(memory_blocks)
        return blocks


# --- Policy ---


class UserRole(str, Enum):
    READER = "reader"
    WRITER = "writer"
    ADMIN = "admin"


@dataclass
class User:
    id: str
    role: UserRole
    tools: list[str]


class UserStore(Protocol):
    def get_user(self, user_id: str) -> User: ...


@dataclass
class PolicyEngine:
    users: UserStore

    def allow(self, req: RunRequest, call: ToolCall) -> RiskLevel:
        user = self.users.get_user(req.user_id)

        if not _contains(user.tools, call.name):
            raise PermissionError(f"tool not allowed for user: {call.name}")

        if call.name == "search_docs":
            return RiskLevel.LOW

        if call.name in ("send_email", "delete_file", "run_shell"):
            if user.role != UserRole.ADMIN:
                raise PermissionError("high-risk tool requires admin role")
            return RiskLevel.HIGH

        raise PermissionError("unknown tool")


# --- Schema validation ---


@dataclass
class ToolSchemaValidator:
    allowed_args: dict[str, list[str]]

    def validate(self, call: ToolCall) -> None:
        allowed = self.allowed_args.get(call.name)
        if allowed is None:
            raise ValueError(f"schema not found for tool: {call.name}")

        for key in call.args:
            if not _contains(allowed, key):
                raise ValueError(f'unexpected arg "{key}" for tool {call.name}')

        if call.name == "search_docs":
            query = call.args.get("query")
            if not isinstance(query, str) or not query:
                raise ValueError("search_docs.query is required")


# --- Safe tool registry ---


class Tool(ABC):
    @abstractmethod
    def call(self, args: dict[str, Any]) -> str:
        raise NotImplementedError


@dataclass
class Registry:
    tools: dict[str, Tool]

    def get(self, name: str) -> Tool:
        tool = self.tools.get(name)
        if tool is None:
            raise KeyError(f"tool not registered: {name}")
        return tool


@dataclass
class Executor:
    registry: Registry
    audit: AuditLogger

    def execute(self, call: ToolCall) -> str:
        tool = self.registry.get(call.name)
        return tool.call(call.args)


# --- Output validation ---


SECRET_PATTERNS = [
    re.compile(r"(?i)api[_-]?key\s*[:=]\s*['\"]?[^'\"\s]+"),
    re.compile(r"(?i)bearer\s+[a-z0-9._\-]+"),
]


@dataclass
class OutputValidatorImpl:
    def validate(self, text: str) -> None:
        for pattern in SECRET_PATTERNS:
            if pattern.search(text):
                raise ValueError("output contains possible secret")


# --- Минимальная сборка runtime ---


def new_runtime(
    input_guard: InputGuard,
    context_builder: ContextBuilder,
    planner: Planner,
    policy: Policy,
    schema: SchemaValidator,
    approval: ApprovalGate,
    executor: ToolExecutor,
    output: OutputValidator,
    audit: AuditLogger,
) -> Runtime:
    return Runtime(
        input_guard=input_guard,
        context_builder=context_builder,
        planner=planner,
        policy=policy,
        schema=schema,
        approval=approval,
        executor=executor,
        output=output,
        audit=audit,
        budget_factory=lambda: Budget(max_steps=8, max_tool_calls=4, max_denied=2),
    )


def _contains(items: list[str], want: str) -> bool:
    return want in items

File: examples/typescript/part-8/24-end-to-end-secure-agent-go.ts
// Illustrative examples for notes/part-8-practice/24-end-to-end-secure-agent-go.md
// Not for production use. Licensed under MIT (see LICENSE-CODE).

enum RiskLevel {
  Low = "Low",
  Medium = "Medium",
  High = "High",
  Critical = "Critical",
}

enum TrustLevel {
  TrustedRuntime = "trusted_runtime",
  UntrustedUser = "untrusted_user",
  UntrustedTool = "untrusted_tool",
}

interface ToolCall {
  name: string;
  args: Record<string, unknown>;
}

interface Plan {
  finalAnswer?: string;
  toolCall?: ToolCall | null;
  reason?: string;
}

interface ContextBlock {
  source: string;
  trust: TrustLevel;
  text: string;
}

interface RunRequest {
  runId: string;
  userId: string;
  task: string;
}

interface RunResult {
  runId: string;
  finalAnswer?: string;
  toolCalls: ToolCall[];
  blocked: string[];
}

interface InputGuard {
  check(task: string): void;
}

interface ContextBuilder {
  build(req: RunRequest): ContextBlock[];
}

interface Planner {
  plan(req: RunRequest, blocks: ContextBlock[]): Plan;
}

interface Policy {
  allow(req: RunRequest, call: ToolCall): RiskLevel;
}

interface SchemaValidator {
  validate(call: ToolCall): void;
}

interface ApprovalGate {
  approve(req: RunRequest, call: ToolCall, risk: RiskLevel): void;
}

interface ToolExecutor {
  execute(call: ToolCall): string;
}

interface OutputValidator {
  validate(text: string): void;
}

interface AuditLogger {
  log(event: AuditEvent): void;
}

interface AuditEvent {
  time: Date;
  runId: string;
  userId?: string;
  event: string;
  component: string;
  tool?: string;
  risk?: RiskLevel;
  decision?: string;
  reason?: string;
  attrs?: Record<string, unknown>;
}

class Budget {
  maxSteps: number;
  maxToolCalls: number;
  maxDenied: number;
  steps = 0;
  toolCalls = 0;
  denied = 0;

  constructor(maxSteps: number, maxToolCalls: number, maxDenied: number) {
    this.maxSteps = maxSteps;
    this.maxToolCalls = maxToolCalls;
    this.maxDenied = maxDenied;
  }

  check(): void {
    if (this.steps > this.maxSteps) {
      throw new Error("budget exceeded: max steps");
    }
    if (this.toolCalls > this.maxToolCalls) {
      throw new Error("budget exceeded: max tool calls");
    }
    if (this.denied > this.maxDenied) {
      throw new Error("budget exceeded: max denied actions");
    }
  }
}

class Runtime {
  inputGuard: InputGuard;
  contextBuilder: ContextBuilder;
  planner: Planner;
  policy: Policy;
  schema: SchemaValidator;
  approval: ApprovalGate;
  executor: ToolExecutor;
  output: OutputValidator;
  audit: AuditLogger;
  budgetFactory: () => Budget;

  constructor(
    inputGuard: InputGuard,
    contextBuilder: ContextBuilder,
    planner: Planner,
    policy: Policy,
    schema: SchemaValidator,
    approval: ApprovalGate,
    executor: ToolExecutor,
    output: OutputValidator,
    audit: AuditLogger,
    budgetFactory: () => Budget,
  ) {
    this.inputGuard = inputGuard;
    this.contextBuilder = contextBuilder;
    this.planner = planner;
    this.policy = policy;
    this.schema = schema;
    this.approval = approval;
    this.executor = executor;
    this.output = output;
    this.audit = audit;
    this.budgetFactory = budgetFactory;
  }

  run(req: RunRequest): RunResult {
    if (!req.runId) {
      throw new Error("run_id is required");
    }

    const result: RunResult = { runId: req.runId, toolCalls: [], blocked: [] };
    const budget = this.budgetFactory();

    try {
      this.inputGuard.check(req.task);
    } catch (err) {
      this.audit.log({
        time: new Date(),
        runId: req.runId,
        userId: req.userId,
        event: "input_blocked",
        component: "input_guard",
        decision: "denied",
        reason: err instanceof Error ? err.message : String(err),
      });
      throw err;
    }

    let blocks = this.contextBuilder.build(req);

    while (true) {
      budget.steps += 1;
      try {
        budget.check();
      } catch (err) {
        result.blocked.push("budget_exceeded");
        throw err;
      }

      const plan = this.planner.plan(req, blocks);

      if (!plan.toolCall) {
        try {
          this.output.validate(plan.finalAnswer ?? "");
        } catch (err) {
          result.blocked.push("output_blocked");
          throw err;
        }

        result.finalAnswer = plan.finalAnswer ?? "";
        return result;
      }

      const call = plan.toolCall;
      budget.toolCalls += 1;

      let risk: RiskLevel;
      try {
        risk = this.policy.allow(req, call);
      } catch (err) {
        budget.denied += 1;
        result.blocked.push(`tool_denied:${call.name}`);
        this.audit.log({
          time: new Date(),
          runId: req.runId,
          userId: req.userId,
          event: "tool_denied",
          component: "policy",
          tool: call.name,
          decision: "denied",
          reason: err instanceof Error ? err.message : String(err),
        });
        throw err;
      }

      try {
        this.schema.validate(call);
      } catch (err) {
        budget.denied += 1;
        result.blocked.push(`schema_validation_failed:${call.name}`);
        throw err;
      }

      if (requiresApproval(risk)) {
        try {
          this.approval.approve(req, call, risk);
        } catch (err) {
          budget.denied += 1;
          result.blocked.push(`approval_rejected:${call.name}`);
          throw err;
        }
      }

      const observation = this.executor.execute(call);
      result.toolCalls.push(call);

      blocks = [
        ...blocks,
        {
          source: `tool:${call.name}`,
          trust: TrustLevel.UntrustedTool,
          text: observation,
        },
      ];
    }
  }
}

function requiresApproval(risk: RiskLevel): boolean {
  return risk === RiskLevel.High || risk === RiskLevel.Critical;
}

class Guard implements InputGuard {
  maxChars: number;

  constructor(maxChars: number) {
    this.maxChars = maxChars;
  }

  check(task: string): void {
    if (!task.trim()) {
      throw new Error("empty task");
    }

    if (task.length > this.maxChars) {
      throw new Error("input too large");
    }

    const lower = task.toLowerCase();
    const suspicious = [
      "ignore previous instructions",
      "disregard system prompt",
      "reveal your system prompt",
      "send secrets",
      "exfiltrate",
    ];

    for (const marker of suspicious) {
      if (lower.includes(marker)) {
        throw new Error("possible prompt injection");
      }
    }
  }
}

interface Memory {
  readForUser(userId: string): ContextBlock[];
}

class Builder implements ContextBuilder {
  memory: Memory;

  constructor(memory: Memory) {
    this.memory = memory;
  }

  build(req: RunRequest): ContextBlock[] {
    const memoryBlocks = this.memory.readForUser(req.userId);
    return [
      {
        source: "user_task",
        trust: TrustLevel.UntrustedUser,
        text: req.task,
      },
      ...memoryBlocks,
    ];
  }
}

enum UserRole {
  Reader = "reader",
  Writer = "writer",
  Admin = "admin",
}

interface User {
  id: string;
  role: UserRole;
  tools: string[];
}

interface UserStore {
  getUser(userId: string): User;
}

class PolicyEngine implements Policy {
  users: UserStore;

  constructor(users: UserStore) {
    this.users = users;
  }

  allow(req: RunRequest, call: ToolCall): RiskLevel {
    const user = this.users.getUser(req.userId);

    if (!contains(user.tools, call.name)) {
      throw new Error(`tool not allowed for user: ${call.name}`);
    }

    if (call.name === "search_docs") {
      return RiskLevel.Low;
    }

    if (call.name === "send_email" || call.name === "delete_file" || call.name === "run_shell") {
      if (user.role !== UserRole.Admin) {
        throw new Error("high-risk tool requires admin role");
      }
      return RiskLevel.High;
    }

    throw new Error("unknown tool");
  }
}

class ToolSchemaValidator implements SchemaValidator {
  allowedArgs: Record<string, string[]>;

  constructor(allowedArgs: Record<string, string[]>) {
    this.allowedArgs = allowedArgs;
  }

  validate(call: ToolCall): void {
    const allowed = this.allowedArgs[call.name];
    if (!allowed) {
      throw new Error(`schema not found for tool: ${call.name}`);
    }

    for (const key of Object.keys(call.args)) {
      if (!contains(allowed, key)) {
        throw new Error(`unexpected arg "${key}" for tool ${call.name}`);
      }
    }

    if (call.name === "search_docs") {
      const query = call.args.query;
      if (typeof query !== "string" || !query) {
        throw new Error("search_docs.query is required");
      }
    }
  }
}

interface Tool {
  call(args: Record<string, unknown>): string;
}

class Registry {
  tools: Record<string, Tool>;

  constructor(tools: Record<string, Tool>) {
    this.tools = tools;
  }

  get(name: string): Tool {
    const tool = this.tools[name];
    if (!tool) {
      throw new Error(`tool not registered: ${name}`);
    }
    return tool;
  }
}

class Executor implements ToolExecutor {
  registry: Registry;
  audit: AuditLogger;

  constructor(registry: Registry, audit: AuditLogger) {
    this.registry = registry;
    this.audit = audit;
  }

  execute(call: ToolCall): string {
    const tool = this.registry.get(call.name);
    return tool.call(call.args);
  }
}

const secretPatterns = [
  /api[_-]?key\s*[:=]\s*['"]?[^'"\s]+/i,
  /bearer\s+[a-z0-9._\-]+/i,
];

class OutputValidatorImpl implements OutputValidator {
  validate(text: string): void {
    for (const pattern of secretPatterns) {
      if (pattern.test(text)) {
        throw new Error("output contains possible secret");
      }
    }
  }
}

function newRuntime(
  inputGuard: InputGuard,
  contextBuilder: ContextBuilder,
  planner: Planner,
  policy: Policy,
  schema: SchemaValidator,
  approval: ApprovalGate,
  executor: ToolExecutor,
  output: OutputValidator,
  audit: AuditLogger,
): Runtime {
  return new Runtime(
    inputGuard,
    contextBuilder,
    planner,
    policy,
    schema,
    approval,
    executor,
    output,
    audit,
    () => new Budget(8, 4, 2),
  );
}

function contains(items: string[], want: string): boolean {
  return items.includes(want);
}
